src/zope/container/ordered.py

```diff
@@ -272,17 +273,14 @@ def updateOrder(self, order):
         if len(order) != len(self._order):
             raise ValueError("Incompatible key set.")
 
-        was_dict = {}
-        will_be_dict = {}
-        new_order = PersistentList()
+        order = [checkAndConvertName(x) for x in order]
 
-        for i, obj in enumerate(order):
-            was_dict[self._order[i]] = 1
-            will_be_dict[obj] = 1
-            new_order.append(obj)
-
-        if will_be_dict != was_dict:
+        if frozenset(order) != frozenset(self._order):
             raise ValueError("Incompatible key set.")
 
-        self._order = new_order
+        if order == self._order:
+            return
+
+        self._order.sort(key=order.index)
```

Comment:
Seems too clever, and adds a quadratic time operation.  If replacing the list is an issue, why not slice assign? e.g.:

```python
    self._order[:] = order
```